components/merkledb/src/access/extensions.rs

```diff
@@ -115,6 +116,40 @@ pub trait AccessExt: Access {
         ProofMapIndex::from_access(self, addr.into()).unwrap()
     }
 
+    /// Generic variant of the proof map. Requires implicit `KeyMode` to be constructed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum_merkledb::{access::AccessExt, Fork, Database, ListIndex, TemporaryDB, ProofMapIndex,
+    /// proof_map_index::{Raw, Hashed}};
+    /// use exonum_crypto::PublicKey;
+    ///
+    /// let db = TemporaryDB::new();
+    /// let fork = db.fork();
+    ///
+    /// // Hashed variant for keys implementing `ObjectHash`.
+    /// let hashed_map: ProofMapIndex<&Fork, u32, u32, Hashed> = fork.get_gen_proof_map("hashed");
+    ///
+    /// // Raw variant for keys that maps directly to `ProofPath`.
+    /// let raw_map: ProofMapIndex<&Fork, PublicKey, u32, Raw> = fork.get_gen_proof_map("raw");
+    /// ```
+    /// # Panics
+    ///
+    /// If the index exists, but is not a Merkelized map.
+    fn get_gen_proof_map<I, K, V, KeyMode>(
```

Comment:
Maybe name it `get_generic_proof_map`?

Reply:
And, by the way, do we need it at all? Maybe it'll be better to have 2 methods: `get_proof_map` and `get_raw_proof_map`?

Reply:
Method `get_raw_proof_map` has been used only once in tests. And also in new `schema` implementations we doesn't use these methods and use `FromAccess::from_acces` instead.

Reply:
Well, we [have](https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/blockchain/schema.rs) places where `get_index_type` methods are used (and generally the same will apply to any situation when eager initialization doesn't fit ones needs), so IMHO it'll be better to have `get_*` methods for any type of index, just to be consistent.

Reply:
We have a method `get_proof_map` (and I don't think that it should be removed), so `get_generic_proof_map` is actually needed only for a raw variant.

Reply:
`get_generic_proof_map` is also needed for tests and that why I've created `TestsExt` trait, but it seems to be a bad idea.

Reply:
The request to refactor the code was mine, so I'll chime in.

> `get_generic_proof_map` is actually needed only for a raw variant.

Not really, since there are no restrictions on implementing custom key styles externally (should there be ones?). I'd argue that schema-like code will either use `FromAccess` (through a derive macro once it's implemented), or have code like this:

```rust
struct Schema<T>(T);

impl<T: Access> Schema<T> {
    pub fn some_map(&self) -> RawProofMapIndex<T::Base, PublicKey, u64> {
        self.0.clone().get_generic_proof_map("some_map")
        // the return type is known, so there's no difference between
        // `generic` and `raw` indexes from the UX perspective.
    }
}
```

Reply:
Not sure if I understand. Why it can't be `self.0.clone().get_raw_proof_map("some_map")` if the return type is already `RawProofMapIndex`?